runner/src/main.rs

```diff
@@ -667,25 +655,52 @@ fn run_example(opt: &RunExamples, example: &Example) -> Step {
     // clients in the foreground.
     #[allow(clippy::collapsible_if)]
     let run_backend_server_clients: Step = if opt.run_server.unwrap_or(true) {
-        let run_server_clients = if opt.build_client.client_variant != NO_CLIENTS {
-            Step::WithBackground {
-                name: "background server".to_string(),
-                background: run_server,
-                foreground: Box::new(run_clients),
+        let run_server_clients = if example.applications.is_empty() {
+            if opt.build_client.client_variant != NO_CLIENTS {
+                run_clients
+            } else {
+                panic!("`NO_CLIENTS` variant is not supported when no applications are provided");
             }
         } else {
-            Step::Single {
-                name: "run server".to_string(),
-                command: run_server,
+            let application = example
+                .applications
+                .get(opt.application_variant.as_str())
+                .expect("Unsupported application variant");
+
+            let run_server = run_example_server(
+                &opt.build_server,
+                &example.server,
+                opt.server_additional_args.clone(),
+                &application.out,
+                &opt.permissions_file,
+            );
+
+            if opt.build_client.client_variant != NO_CLIENTS {
+                Step::WithBackground {
+                    name: "background server".to_string(),
+                    background: run_server,
+                    foreground: Box::new(run_clients),
+                }
+            } else {
+                Step::Single {
+                    name: "run server".to_string(),
+                    command: run_server,
+                }
             }
         };
-        match &example.backend {
-            Some(backend) => Step::WithBackground {
-                name: "background backend".to_string(),
-                background: run(&backend, &opt.build_client, Vec::new()),
-                foreground: Box::new(run_server_clients),
-            },
-            None => run_server_clients,
+        if example.backends.is_empty() {
+            run_server_clients
+        } else {
```

Comment:
Why does this need special treatment? Ideally the fold should cover the empty case just fine?